Beautiful console report

gucong3000 opened the pull request:
Fixes #49 

![image](https://cloud.githubusercontent.com/assets/1171691/23452899/0ce7b5f0-fea1-11e6-9dd2-dd12f1e5a47a.png)

- [x] new API for `Error` object
- [x] modify test case for new API
- [x] remove support for node v0.x
- [x] add test case for [eclint.ts](eclint.ts)
- [x] use `gulp-reporter` to show report

jedmao commented on .vscode/settings.json:
```diff
@@ -0,0 +1,3 @@
+{
```
Can you get rid of this file or anything editor-specific?

jedmao commented on lib/eclint.spec.ts:
```diff
@@ -1,6 +1,223 @@
+				expect(file.editorconfig.fixed).to.be.ok;
+				expect(file.editorconfig.errors).to.have.lengthOf(0);
+			}).on('end', () => {
+				done();
```
I believe you should only require one `done` in a test. Is `done` ever called twice here? If not, which one is being called? That should be the only `done` in the test. Please do the same throughout the PR.

jedmao commented on lib/eclint.spec.ts:
```diff
@@ -1,6 +1,223 @@
+			}).on('error', done);
+		});
+
+		it('check after fix', (done) => {
```
checks (plural)

jedmao commented on lib/eclint.spec.ts:
```diff
@@ -1,6 +1,223 @@
+			});
+		});
+
+		it('fix after check', (done) => {
```
"fixes" after check

jedmao commented on lib/eclint.spec.ts:
```diff
@@ -1,6 +1,223 @@
+
+	describe('end_of_line rule', () => {
+
+		it('invalid newline: crlf, expected: lf', (done) => {
```
`it('throws invalid...` please.

jedmao commented on lib/eclint.spec.ts:
```diff
@@ -1,6 +1,223 @@
+
+	describe('insert_final_newline rule', () => {
+
+		it('expected final newline', (done) => {
```
"expects"

jedmao commented on lib/eclint.spec.ts:
```diff
@@ -1,6 +1,223 @@
+				expect(error.name).to.equal('EditorConfigError');
+				expect(error.message).to.equal('unexpected final newline');
+				expect(error.rule).to.equal('insert_final_newline');
+				// expect(error.source).to.equal('');
```
Accident? Please remove comments.

jedmao commented on lib/eclint.ts:
```diff
@@ -179,18 +196,28 @@ module eclint {
 					});
 
+					updateResult(file, {
+						fixed: Boolean(file.editorconfig && file.editorconfig.fixed),
```
`new Boolean` please.

jedmao commented on lib/editor-config-error.ts:
```diff
@@ -0,0 +1,22 @@
+class EditorConfigError extends Error {
+	fileName: string = '<anonymous>';
```
Nice!

jedmao commented on lib/editor-config-error.ts:
```diff
@@ -0,0 +1,22 @@
+	message: string = '';
+	rule: string = '';
+	source: string = '';
+	name: string = 'EditorConfigError';
```
I think you can safely remove the types here in this section, as they will be inferred for you. Example:
```js
source = '';
name = 'EditorConfigError';
```

jedmao commented on lib/editor-config-error.ts:
```diff
@@ -0,0 +1,22 @@
+		return [
+			`${ this.name }: ${ this.message } (${ this.rule })`,
+			`    at (${ this.fileName }:${ this.lineNumber }:${ this.columnNumber })`
+		].join('\n');
```
Did you consider using `os.EOL`?

jedmao commented on lib/rules/indent_size.ts:
```diff
@@ -33,13 +34,17 @@ function check(context: eclint.Context, settings: eclint.Settings, doc: linez.Do
 				'expected: ' + configSetting
 			].join(' '));
+			error.lineNumber = line.number;
+			// error.columnNumber = 1;
```
Remove or fix comment please.

jedmao commented on lib/rules/indent_style.ts:
```diff
@@ -16,40 +17,39 @@ function resolve(settings: eclint.Settings) {
-					'invalid indentation: found a leading space, expected: tab'
-				].join(' '));
-				return;
+				return creatErroe('invalid indentation: found a leading space, expected: tab');
```
Looks like a typo "createErroe"?

jedmao commented on lib/rules/indent_style.ts:
```diff
@@ -16,40 +17,39 @@ function resolve(settings: eclint.Settings) {
-					'invalid indentation: found ' + softTabCount + ' soft tab'
-				].join(' ') + ((softTabCount > 1) ? 's' : ''));
-				return;
+				return creatErroe('invalid indentation: found ' + softTabCount + ' soft tab');
```
same typo

jedmao replied:
Please use a string template for this:
```ts
return createError(`invalid indentation: found ${softTabCount} soft tab(s)');
```
Notice I said "tab(s)", as the number can be singular or plural.

jedmao commented on lib/rules/indent_style.ts:
```diff
@@ -16,40 +17,39 @@ function resolve(settings: eclint.Settings) {
-					'invalid indentation: found ' + hardTabCount + ' hard tab'
-				].join(' ') + ((hardTabCount > 1) ? 's' : ''));
-				return;
+				return creatErroe('invalid indentation: found ' + hardTabCount + ' hard tab');
```
same as https://github.com/jedmao/eclint/pull/52/files#r103755967

jedmao commented on .gitignore:
```diff
@@ -1,3 +1,4 @@
 coverage/
 node_modules/
 js/
+.vscode/
```
This should go in your [global `.gitignore`](https://help.github.com/articles/ignoring-files/#create-a-global-gitignore).

gucong3000 replied:
Removed.

jedmao commented on lib/eclint.spec.ts:
```diff
@@ -141,7 +141,7 @@ describe('eclint gulp plugin', () => {
 	describe('end_of_line rule', () => {
 
-		it('invalid newline: crlf, expected: lf', (done) => {
+		it('throws invalid "invalid newline: crlf, expected: lf"', (done) => {
```
please remove the first "invalid"

jedmao commented on lib/rules/indent_style.ts:
```diff
@@ -18,13 +18,7 @@ function resolve(settings: eclint.Settings) {
-			}
-			return str;
-		});
+	function createError(message: string, columnNumber: number = 1) {
```
Sorry, I should have caught this sooner, but you also don't need to declare a type here, as it will be inferred by your default value of `1` to be a `number`. Example:
```ts
function foo(bar = 1) {
  return bar; // number
}
```

jedmao commented on d.ts/eclint.d.ts:
```diff
@@ -2,6 +2,8 @@
 /// <reference path="../typings/gulp-util/gulp-util.d.ts" />
 /// <reference path="../node_modules/linez/linez.d.ts" />
```
If you upgrade `linez` to the latest version and change the import to `import linez from 'linez';` you shouldn't need this `<reference` any more.

jedmao commented on d.ts/eclint.d.ts:
```diff
@@ -2,6 +2,8 @@
 /// <reference path="../typings/gulp-util/gulp-util.d.ts" />
```
If TypeScript is updated, you can add this path to the `includes` section of `tsconfig.json` and you won't need this either. If you don't want to make this update, that's fine. I might go over it like I did with `linez` and fix it up later myself.

jedmao commented on lib/eclint.ts:
```diff
@@ -179,18 +196,29 @@ module eclint {
 					});
 
+					updateResult(file, {
+						// Forced type conversion, not wrapper class
```
please remove the comment.

jedmao commented on lib/eclint.ts:
```diff
@@ -179,18 +196,29 @@ module eclint {
 
+					updateResult(file, {
+						// Forced type conversion, not wrapper class
+						fixed: Boolean(file.editorconfig && file.editorconfig.fixed),
```
Have you considered `!!A` in cases like these?
```ts
fixed: !!(file.editorconfig && file.editorconfig.fixed),
```
Also, you can use `lodash/get` here like so:
```ts
import { get } from 'lodash';
fixed: !!get(file, 'editorconfig.fixed'),
```

jedmao commented:
Why remove the coverage threshold and check coverage? You can just adjust the numbers to whatever it is today and ensure no future code is introduced w/o coverage.

gucong3000 commented:
Everything is ready to merge.

jedmao commented:
Can you squash these commits into one and give it a meaningful commit messages? Also, can you give me an outline of all new features introduced, fixes and breaking changes, so I can update the changelog accordingly?

jedmao commented:
Maybe the checklist you provided above is good enough for the outline. Just make sure it's updated please.

gucong3000 commented:
- Dropped: Node 0.12 support.
- Added: support for Node 4+.
- Added: use `gulp-reporter` for beautiful console report in CLI. [#49](https://github.com/jedmao/eclint/issues/49)
- Added: class `EditorConfigError`.
- Added: interface `file.editorconfig: EditorConfigLintResult`, for compatibility with `gulp-reporter`.
- Modified: interface for `LineRule.check()`, returns `EditorConfigError`.
- Modified: interface for `DocumentRule.check()`, returns `EditorConfigError[]`
- Modified: Upgraded all dependencies, except `lodash`.
- Fixed: error `TS2307: Cannot find module 'd.ts/linez'.` [#38](https://github.com/jedmao/eclint/issues/38).
- Fixed: interface definition `CheckCommandOptions.reporter(file: EditorConfigLintFile, error: EditorConfigError)`, ensure consistent with practice.
- Modified: test case modified for `Rule.check()`.

**Thank you for your patience in these days.  Pandas wish you good health **

jedmao commented:
Closed via #53. Is there anything else you want to do before we do a major release?
